modules/gapi/src/streaming/onevpl/accelerators/surface/cpu_frame_adapter.cpp

```diff
@@ -0,0 +1,121 @@
+    parent_surface_ptr(surface) {
+
+    GAPI_Assert(parent_surface_ptr && "Surface is nullptr");
+    parent_surface_ptr->obtain_lock();
+
+
+    const Surface::info_t& info = parent_surface_ptr->get_info();
+    const Surface::data_t& data = parent_surface_ptr->get_data();
+
+    GAPI_LOG_DEBUG(nullptr, "surface: " << parent_surface_ptr->get_handle() <<
+                            ", w: " << info.Width << ", h: " << info.Height <<
+                            ", p: " << data.Pitch);
+}
+
+VPLMediaFrameCPUAdapter::~VPLMediaFrameCPUAdapter() {
+
+    // Each VPLMediaFrameCPUAdapter releases mfx surface counter
+    // The last VPLMediaFrameCPUAdapter releases shared Surface pointer
+    // The last surface pointer releases workspace memory
+    parent_surface_ptr->release_lock();
+}
+
+cv::GFrameDesc VPLMediaFrameCPUAdapter::meta() const {
+    GFrameDesc desc;
+    const Surface::info_t& info = parent_surface_ptr->get_info();
+    switch(info.FourCC)
+    {
+        case MFX_FOURCC_I420:
+            throw std::runtime_error("MediaFrame doesn't support I420 type");
+            break;
+        case MFX_FOURCC_NV12:
+            desc.fmt = MediaFormat::NV12;
+            break;
+        default:
+            throw std::runtime_error("MediaFrame unknown 'fmt' type: " + std::to_string(info.FourCC));
+    }
+
+    desc.size = cv::Size{info.Width, info.Height};
+    return desc;
+}
+
+MediaFrame::View VPLMediaFrameCPUAdapter::access(MediaFrame::Access) {
+    const Surface::data_t& data = parent_surface_ptr->get_data();
+    const Surface::info_t& info = parent_surface_ptr->get_info();
+    using stride_t = typename cv::MediaFrame::View::Strides::value_type;
+    GAPI_Assert(data.Pitch >= 0 && "Pitch is less than 0");
+
+    stride_t pitch = static_cast<stride_t>(data.Pitch);
+    switch(info.FourCC) {
+        case MFX_FOURCC_I420:
+        {
+            cv::MediaFrame::View::Ptrs pp = {
+                data.Y,
+                data.U,
+                data.V,
+                nullptr
+                };
+            cv::MediaFrame::View::Strides ss = {
+                    pitch,
+                    pitch / 2,
+                    pitch / 2, 0u
+                };
+            return cv::MediaFrame::View(std::move(pp), std::move(ss));
+        }
+        case MFX_FOURCC_NV12:
+        {
+            cv::MediaFrame::View::Ptrs pp = {
+                data.Y,
+                data.UV, nullptr, nullptr
+                };
+            cv::MediaFrame::View::Strides ss = {
+                    pitch,
+                    pitch, 0u, 0u
+                };
+            return cv::MediaFrame::View(std::move(pp), std::move(ss));
+        }
+            break;
+        default:
+            throw std::runtime_error("MediaFrame unknown 'fmt' type: " + std::to_string(info.FourCC));
+    }
+}
+
+cv::util::any VPLMediaFrameCPUAdapter::blobParams() const {
+    GAPI_Assert("VPLMediaFrameCPUAdapter::blobParams() is not implemented");
```

Comment:
Maybe it is worth to introduce a macro for that (where method name is taken automatically via preprocessor)

Reply:
found in `cvdefs.h`

```
#ifdef CV_Func
// keep current value (through OpenCV port file)
#elif defined __GNUC__ || (defined (__cpluscplus) && (__cpluscplus >= 201103))
#define CV_Func __func__
#elif defined __clang__ && (__clang_minor__ * 100 + __clang_major__ >= 305)
#define CV_Func __func__
#elif defined(__STDC_VERSION__) && (__STDC_VERSION >= 199901)
#define CV_Func __func__
#elif defined _MSC_VER
#define CV_Func __FUNCTION__
#elif defined(__INTEL_COMPILER) && (_INTEL_COMPILER >= 600)
#define CV_Func __FUNCTION__
#elif defined __IBMCPP__ && __IBMCPP__ >=500
#define CV_Func __FUNCTION__
#elif defined __BORLAND__ && (__BORLANDC__ >= 0x550)
#define CV_Func __FUNC__
#else
#define CV_Func "<unknown>"
#endif
```

i'll change it on 
```
GAPI_Assert(false && CV_Func
                         "is not implemented");
```

Reply:
> CV_Func

Don't forget about G-API standalone mode. `CV_` macros should be used carefully.

OpenCV mode doesn't require that, because function name is a part of exception.

Reply:
thanks, i think i'll revert it back

Reply:
I was thinking of a `GAPI_NOT_IMPLEMENTED()` kind of macro to be honest :)